cc/algorithms/partition-selection-strategies.h

```diff
@@ -0,0 +1,345 @@
+    }
+    else {
+      adjusted_epsilon_ = 0;
+    }
+    crossover_1_ = 1 + floor((1.0/adjusted_epsilon_)
+                     * log((exp(adjusted_epsilon_)
+                     + 2.0 * adjusted_delta_ - 1.0)
+                       / ((exp(adjusted_epsilon_) + 1) * adjusted_delta_)));
+    crossover_2_ = crossover_1_ + floor((1.0/adjusted_epsilon_)
+                                  *  log(1 + ((exp(adjusted_epsilon_) - 1)
+                                  / adjusted_delta_)
+                                  * (1 - ProbabilityOfKeep(crossover_1_))));
+  }
+
+  virtual ~PreaggPartitionSelection() = default;
+
+  double GetAdjustedEpsilon() { return adjusted_epsilon_; }
+
+  double GetFirstCrossover() { return crossover_1_; }
+
+  double GetSecondCrossover() { return crossover_2_; }
+
+  bool ShouldKeep(int num_users) override {
+    //generate a random number between 0 and 1
+    double rand_num = UniformDouble();
+    //only keep partition if random number < expected probability of keep
+    return (rand_num <= ProbabilityOfKeep(num_users));
+  }
+
+ protected:
+  double adjusted_epsilon_;
+  double crossover_1_;
+  double crossover_2_;
+
+  double ProbabilityOfKeep(double n) {
+    if(n == 0) {
+      return 0;
+    }
+    else if (n <= crossover_1_) {
+      return (((exp(n * adjusted_epsilon_) - 1) / (exp(adjusted_epsilon_) - 1))
+              * adjusted_delta_);
+    }
+    else if (n > crossover_1_ && n <= crossover_2_) {
+      double m = n - crossover_1_;
+      return ((1 - exp(-1 * m * adjusted_epsilon_))
+               * (1 + adjusted_delta_ / (exp(adjusted_epsilon_) - 1))
+               + exp(-1 * m * adjusted_epsilon_)
+               * ProbabilityOfKeep(crossover_1_));
+    }
+    else {
+      return 1;
+    }
+  }
+
+};
+
+
+class LaplacePartitionSelection : public PartitionSelectionStrategy {
+ public:
+  //Builder for LaplacePartitionSelection
+  class Builder : public PartitionSelectionStrategy::Builder {
+   public:
+     Builder& SetLaplaceMechanism(LaplaceMechanism::Builder* laplace_builder) {
+       laplace_builder_ = laplace_builder;
+       return *static_cast<Builder*>(this);
+     }
+
+     base::StatusOr<std::unique_ptr<PartitionSelectionStrategy>>
+     Build() override {
+       base::Status max_partitions_contributed_status =
+         MaxPartitionsContributedIsSetAndValid();
+       base::Status delta_status = DeltaIsSetAndValid();
+       if (!max_partitions_contributed_status.ok()) {
+         return max_partitions_contributed_status;
+       }
+       else if (!epsilon_.has_value()) {
+         return base::InvalidArgumentError("Epsilon has to be set.");
+       }
+       else if (!delta_status.ok()) {
+         return delta_status;
+       }
+       else if (laplace_builder_.has_value()) {
+         std::unique_ptr<NumericalMechanism> mechanism_ =
+           (*(laplace_builder_.value())).SetEpsilon(epsilon_.value())
+           .SetL1Sensitivity(max_partitions_contributed_.value()).Build()
```

Comment:
nit: I think it's a little clearer to use SetL0Sensitivity and SetLinfSensitivity here instead of SetL1Sensitivity, since L0 sensitivity maps directly to max_partitions_contributed_.